packages/yew-router/src/hidden.rs

```diff
@@ -0,0 +1,407 @@
+        if let Arg::Opaque(v) = value {
+            Ok(v)
+        } else {
+            Err(())
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Args {
+    args: Vec<Arg>,
+    route: Route,
+}
+
+impl Args {
+    pub fn new(route: Route) -> Self {
+        Self {
+            route,
+            args: Vec::new(),
+        }
+    }
+    pub fn empty() -> Self {
+        Self::new(Route::new("/"))
+    }
+    pub fn push<T: Into<Arg>>(&mut self, arg: T) {
+        self.args.push(arg.into());
+    }
+    pub fn pop<T>(&mut self) -> Option<T>
+    where
+        Arg: TryInto<T>,
+    {
+        self.args.pop().expect("No argument to pop").try_into().ok()
+    }
+    pub fn push_serde<T: Serialize>(&mut self, arg: &T) {
+        self.push(JsValue::from_serde(arg).expect("Argument to serialize"));
+    }
+    pub fn pop_serde<T: DeserializeOwned>(&mut self) -> Option<T> {
+        self.pop::<JsValue>()?.into_serde().ok()
+    }
+    pub fn push_str<T: Display>(&mut self, arg: &T) {
+        self.push(arg.to_string());
+    }
+    pub fn pop_str<T: FromStr>(&mut self) -> Option<T> {
+        self.pop::<String>()?.parse().ok()
+    }
+    pub fn take_route(&mut self) -> Route {
+        mem::replace(&mut self.route, Default::default())
+    }
+    pub fn store_route(&mut self, route: Route) {
+        self.route = route;
+    }
+}
+
+/// An instance of this type is constructed for each variant by the
+/// derive macro.
+pub struct RouteMuncher<T: DerivedRoutable> {
+    pub regex: &'static str,
+    pub munchers: &'static [&'static dyn Muncher],
+    pub ctor: &'static dyn Fn(Args) -> Option<T>,
+}
+
+impl<T: DerivedRoutable> RouteMuncher<T> {
+    /// Extracts a list of arguments from the `Route` using the configured
+    /// munchers.
+    fn munch(&self, route: &Route) -> Option<Args> {
+        let mut args = Args::new(route.clone());
+        for muncher in self.munchers {
+            if !muncher.munch(&mut args) {
+                return None;
+            }
+        }
+        Some(args)
+    }
+    /// Attempts to construct this variant from the provided `Route`.
+    fn from_route(&self, route: &Route) -> Option<T> {
+        (self.ctor)(self.munch(route)?)
+    }
+    /// Builds a route from a list of arguments.
+    fn reconstitute(&self, mut args: Args) -> Route {
+        for muncher in self.munchers.iter().rev() {
+            muncher.reconstitute(&mut args);
+        }
+        args.take_route()
+    }
+}
+
+/// This trait is implemented for each possible way arguments can be
+/// extracted from a route.
+pub trait Muncher {
+    fn munch(&self, args: &mut Args) -> bool;
+    fn reconstitute(&self, args: &mut Args);
+}
+
+// Configuration for the `PathSegmentMuncher`
+pub enum PathPart {
+    Match(&'static str),
+    ExtractOne,
+    ExtractAll,
+}
+
+/// A muncher which matches and/or extracts path segments of a URL.
+pub struct PathSegmentMuncher {
+    pub parts: &'static [PathPart],
+}
```

Comment:
Why parse segments manually when there's a crate (`route_recognizer`) to do that?

Reply:
- `route_recognizer` does both steps as one (recognising the route and then extracting arguments) whereas I need them to be split.
- Its `Router` type needs to be constructed dynamically.
- It returns parsed arguments in a format that would need to be converted to be compatible with other munchers.
- It does not easily allow getting the "rest" of the route without doing some pre-processing step on the input.

Finally, if you look at what this muncher actually does, it's just splitting on `/`, and that functionality is built into the rust std anyway, so why use a separate crate?

Reply:
I think we should move this logic out of this file then. parser logic and "muncher" should be in their own individual modules imo.

Reply:
Sure, they'll need to be submodules but that's fine.

Reply:
Done.